### PR TITLE
Add proposal for junit jupiter testkit API

### PR DESCRIPTION
**Pull request by britter**

### CODE REVIEW

File: src/main/java/com/github/britter/junit/jupiter/testkit/TestExecutionStatistics.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    Statistics getTestStatistics();
+
+    Statistics getContainerStatistics();
+
+    interface Statistics {
+        // TODO does this belong here?
+        long getReportingEntryPublishedCount();
+
+        // TODO does this belong here?
+        long getDynamicTestRegisteredCount();
```
Comment (dotCipher) [inline src/main/java/com/github/britter/junit/jupiter/testkit/TestExecutionStatistics.java]: I would be more inclined to have something like `getEventCount(Type type)` to reduce API pollution

File: src/test/java/com/github/britter/junit/jupiter/testkit/TestkitExamplesTests.java [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+import static com.github.britter.junit.jupiter.testkit.ExecutedElementType.CONTAINER;
+import static com.github.britter.junit.jupiter.testkit.ExecutedElementType.TEST;
+import static com.github.britter.junit.jupiter.testkit.ExecutionEvent.Type.SKIPPED;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.platform.engine.TestExecutionResult.Status.FAILED;
+
+class TestkitExamplesTests {
+
+    @ExtensionTest(InnerTestCase.class)
```
Comment (dotCipher) [inline src/test/java/com/github/britter/junit/jupiter/testkit/TestkitExamplesTests.java]: I think the goal here is that we want to have the testkit have names that aren't tied specifically do the concept of an Extension.  There are usages around meta testing parts of the junit modules themselves (not even related to Extensions) where the testkit is used, so having the name `ExtensionTest` seems a bit misleading?

File: src/test/java/com/github/britter/junit/jupiter/testkit/TestkitExamplesTests.java [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+        assertEquals(1, events.size());
+    }
+
+    @ExtensionTest(value = InnerTestCase.class, methods = "test_method_2")
+    void execute_selected_methods_of_class_filtering_only_failed_tests(@EventsFilter(status = FAILED) List<ExecutionEvent> events) {
+        assertEquals(0, events.size());
+    }
+
+    @ExtensionTest(value = InnerTestCase.class, methods = "test_method_2")
+    void execute_selected_methods_of_class_with_complex_filtering(@EventsFilter(value = ExecutionEvent.Type.FINISHED, type = TEST, status = FAILED) List<ExecutionEvent> events) {
```
Comment (dotCipher) [inline src/test/java/com/github/britter/junit/jupiter/testkit/TestkitExamplesTests.java]: Pretty neat, is the plan here to just auto-capture the events using `@ExtensionTest` and inject based on the parameters `@EventsFilter` etc?

File: src/test/java/com/github/britter/junit/jupiter/testkit/TestkitExamplesTests.java (hunk 2 above)
Comment (britter) [reply src/test/java/com/github/britter/junit/jupiter/testkit/TestkitExamplesTests.java]: You're write. I talked to @sormuras about this and the plan is to do implement this jupiter API in a separate project, maybe in the junit-pioneer organization

File: src/test/java/com/github/britter/junit/jupiter/testkit/TestkitExamplesTests.java (hunk 3 above)
Comment (britter) [reply src/test/java/com/github/britter/junit/jupiter/testkit/TestkitExamplesTests.java]: I like the idea of specifying what you're interested in. But I don't know whether this annotation based filter API is good for evolving the API.

**Pull request closed without merging**